nidx/src/maintenance/scheduler.rs

```diff
@@ -167,37 +170,150 @@ pub async fn purge_deletions(meta: &NidxMetadata, consumer: &mut PullConsumer) -
+/// marks the top bucket, and a configured bucket size marks the next ones.
 ///
-/// Right now, we schedule merges with everything we can
-async fn schedule_merges(meta: &NidxMetadata, last_indexed_seq: Seq) -> anyhow::Result<()> {
-    // TODO: better merge algorithm
-    let merges = sqlx::query!(
-        r#"SELECT index_id, array_agg(id) AS "segment_ids!: Vec<SegmentId>" FROM segments WHERE delete_at IS NULL AND merge_job_id IS NULL AND seq <= $1 GROUP BY index_id"#,
-        i64::from(last_indexed_seq),
-    )
-    .fetch_all(&meta.pool)
-    .await?;
-    for m in merges {
-        if m.segment_ids.len() > 3 {
-            MergeJob::create(meta, m.index_id.into(), &m.segment_ids, last_indexed_seq).await?;
+/// To avoid merging segments too large or dividing in too many buckets for
+/// small segments, one can configure the top and bottom bucket ceilings.
+///
+/// Here's a graphical representation:
+///
+/// ```ignore
+///                              |   (too big to   |
+///                              |    be merged)   |
+///  `top_bucket_max_records` -> +-----------------+
+///                              |  (no segments)  |
+///                              +-----------------+ <- log2(biggest_segment.records)
+///                              |   top bucket    |
+///                              +-----------------+ <- log2(biggest_segment.records) - bucket_size_log
+///                              |                 |
+///                              +-----------------+ <- log2(biggest_segment.records) - 2 * bucket_size_log
+///                              |                 |
+///                              |       ...       |
+///                              |                 |
+///                          +-- +-----------------+
+///       `bucket_size_log` -+   |                 |
+///                          +-- +-----------------+
+///                              |                 |
+///                              +-----------------+ <- log2(biggest_segment.records) - n * bucket_size_log
+/// `bottom_bucket_threshold` -> +  -   -   -  -  -|
+///                              |  bottom bucket  |
+///                              |                 |
+///                              |   (too small    |
+///                              |    to split)    |
+/// ```
+///
+/// TODO: merge segments with too many deletions
+///
+struct MergeScheduler {
+    /// Minimum number of segments needed to perform a merge for an index
+    min_number_of_segments: usize,
+
+    /// Max number of records for a segment to be in the top bucket, i.e.,
+    /// elegible for merge. Once a segment becomes bigger, it won't be merged
+    /// anymore
+    top_bucket_max_records: usize,
+
+    /// Max number of records for a segment to be considered in the bottom
+    /// bucket. Segments with fewer records won't be further splitted in buckets
+    bottom_bucket_threshold: u32,
```

Comment:
It's weird to have multiple integer formats here, maybe standardize everything to usize?